Sources/ApacheBeam/Runtime/DataplaneClient.swift

```diff
@@ -0,0 +1,232 @@
+import Logging
+
+/// Client for handling the multiplexing and demultiplexing of Dataplane messages
+actor DataplaneClient {
+    public struct Pair: Hashable {
+        let id: String
+        let transform: String
+    }
+
+    public enum Message {
+        case data(Data)
+        case timer(String, Data)
+        case last(String, String)
+        case flush
+    }
+
+    struct Multiplex {
+        let id: String
+        let transform: String
+        let message: Message
+    }
+
+    typealias InternalStream = AsyncStream<Multiplex>
+    typealias Stream = AsyncStream<Message>
+
+    public struct MultiplexContinuation {
+        let id: String
+        let transform: String
+        let base: InternalStream.Continuation
+
+        @discardableResult
+        func yield(_ value: Message) -> InternalStream.Continuation.YieldResult {
+            base.yield(Multiplex(id: id, transform: transform, message: value))
+        }
+
+        func finish() {
+            // Does nothing
+        }
+    }
+
+    private let id: String
+    private let log: Logging.Logger
+    private let multiplex: (InternalStream, InternalStream.Continuation)
+    private var streams: [Pair: (Stream, Stream.Continuation, MultiplexContinuation)] = [:]
+    private let flush: Int
+
+    public init(id: String, endpoint: ApiServiceDescriptor, flush: Int = 1000) throws {
+        self.id = id
+        log = Logging.Logger(label: "Dataplane(\(id),\(endpoint.url))")
+        multiplex = AsyncStream.makeStream(of: Multiplex.self)
+        self.flush = flush
+        let client = try Org_Apache_Beam_Model_FnExecution_V1_BeamFnDataAsyncClient(channel: GRPCChannelPool.with(endpoint: endpoint, eventLoopGroup: PlatformSupport.makeEventLoopGroup(loopCount: 1)), defaultCallOptions: CallOptions(customMetadata: ["worker_id": id]))
+        let stream = client.makeDataCall()
+
+        // Mux task
+        Task {
+            log.info("Initiating data plane multiplexing.")
+
+            let input = multiplex.0
+            var count = 0
+            var flushes = 0
+
+            var elements = Org_Apache_Beam_Model_FnExecution_V1_Elements()
+            for try await element in input {
+                var shouldFlush = false
+                switch element.message {
+                case let .data(payload):
+                    elements.data.append(.with {
+                        $0.instructionID = element.id
+                        $0.transformID = element.transform
+                        $0.data = payload
+                    })
+                    count += 1
+                case let .timer(family, payload):
+                    elements.timers.append(.with {
+                        $0.instructionID = element.id
+                        $0.transformID = element.transform
+                        $0.timerFamilyID = family
+                        $0.timers = payload
+                    })
+                    count += 1
+                case let .last(id, transform):
```

Comment:
Two concerns here:

1. Same as before, uf all true cases are executed this should be OK.

2. The Timer sequence and the Data sequence have independent "is_last" signals, https://github.com/apache/beam/blob/master/model/fn-execution/src/main/proto/org/apache/beam/model/fn_execution/v1/beam_fn_api.proto#L735+L761 so it's not clear to me that this will do the right thing. This is because they will be routed to different transforms (Data is always routed to a "datasource" transform, and timers to user transform hosted callbacks).

Reply:
Oh! This is the "writing" side. Probably fine then.